### PR TITLE
Added passive mode to support backup peers

### PR DESCRIPTION
**Pull request by tionis**
This change adds a new optional setting to the joinRoom API to enable "passive mode". Peers in passive mode only ever listen to announce messages but don't announce themselves. This allows them to only be active when normal peers are in the room. With this change it should be feasible to have always-online super peers that join a large amount of rooms without needless connections to e.g. other superpeers. This could allow e.g. offline sync of CRDT data.

This implements https://github.com/dmotz/trystero/issues/146

### CODE REVIEW

File: src/torrent.js [hunk 1]
```diff
@@ -112,7 +113,10 @@ export const joinRoom = strategy({
   subscribe: (client, rootTopic, _, onMessage, getOffers) => {
     const {url} = client
 
-    const announce = async () => {
+    const announce = async (active) => {
```
Comment (dmotz) [inline src/torrent.js]: @tionis Is there a case where `active` should be called with `false`? The only path I can see is via line 89 where the interval gets reset with a new timing and will start calling announce with `undefined`, but I'm not sure if this is intentional or not.

Comment (tionis) [reply src/torrent.js]: Good catch! This is a remnant from when I tried a different approach for the announcing logic before. This bug would stop announcements when the interval is reset. I removed the active parameter and fixed the announcement function handling.

**Comment by tionis** [comment]
Ok, I fixed some logic errors, did some more testing and also added defensive programming style checks. This should be more robust. I'll mark the preview ready for review again.

**Comment by dmotz** [comment]
Thanks, this is improving, and I like that you consolidated most of the logic into strategy.js.

One issue I notice is this scenario:
1. a passive peer receives an announce message, and then announces itself
2. this will cause it to connect to other passive peers if they exist
3. when the last active peer leaves, `checkDeactivate()` will fail because `connectedPeers` still contains the other passive peer(s)
4. this brings us back to zombie swarms

We could address this by sending an additional passive flag in the SDP that the relays exchange, then peers know who's passive before trying to connect, but as noted in the torrent strategy, some trackers deny extra custom fields (maybe we could hide it in another field). You might also think of a better way to do it.

I also don't see why the new try/catch wrapper is beneficial since it swallows errors and logs them.

I support adding this feature, but it needs a bit more work.

**Comment by dmotz** [comment]
@tionis before you dive back in, there is a large refactor of the Trystero codebase in progress on the branch `next2`. I don't know if passive peers will make the next release, but I'd like to support it in the future. I mention this since we'll need to port this passive work over to the new codebase eventually. Feel free to try the new version in your apps and provide feedback.

**Comment by tionis** [comment]
Thanks for the heads-up, I'll take a look at it.

**Comment by tionis** [comment]
This seems to mostly be working, but I'll do some real-world testing for edge cases first, before undrafting it.

**Comment by tionis** [comment]
Ok I've done some testing and most strategies seem to work find in the new tests and manual in browser testing. I still need if connection sharing works and allow efficiently joining thousands of rooms passively.

**Comment by tionis** [comment]
Yeah, having a passive peer with many rooms subscriptions won't work for e.g. nostr as most relays severely limit simultaneous subscriptions. I will look into implementing some kind of (incremental) batching to solve this. Other backends might need similar solutions.

**Comment by tionis** [comment]
Pushed another cleanup pass. (And changed my name in the commits)

This should address the main concerns: passive peers now ignore other passive peers, torrent drops back to dormant `left: 0` announces after deactivation, and Nostr subscriptions are batched/chunked so many passive rooms don’t create tons of relay subs.
I also added/updated tests for passive mode, 1000-room scaling, and Nostr batching.

There are still some probably unfixable Caveats: Firebase RTDB still can’t really batch listeners across rooms, so it’s probably not ideal for huge passive-room counts. Nostr should be much better now, but relay-specific filter limits may still vary.

**Comment by dmotz** [comment]
Thanks for your work on this, just published passive support in 0.25.0. I ended up changing the implementation a bit but it keeps the same functionality and API design.

Let me know if you ever want to share what projects you're using passive peers for.
